src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1156,12 +1159,23 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     nameof(expression));
             }
 
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            if (_placeholderInputTypes.Contains(inputTypeString))
+            {
+                modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+
+                if (modelExplorer.Metadata.Placeholder != null)
```

Comment:
If you mean the ExpressionMetadataProvider bit, I think I'll leave that one alone as I see there are four total such usages in this class and I'm not sure what the team would like to do with that.